File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java
/*
 * Copyright 2015 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire;

import com.squareup.wire.Message.Builder;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Collections;
import java.util.List;
import okio.Buffer;
import okio.BufferedSink;
import okio.BufferedSource;
import okio.ByteString;
import okio.Okio;

import static com.squareup.wire.Preconditions.checkNotNull;
import static com.squareup.wire.ProtoWriter.decodeZigZag32;
import static com.squareup.wire.ProtoWriter.decodeZigZag64;
import static com.squareup.wire.ProtoWriter.encodeZigZag32;
import static com.squareup.wire.ProtoWriter.encodeZigZag64;
import static com.squareup.wire.ProtoWriter.int32Size;
import static com.squareup.wire.ProtoWriter.utf8Length;
import static com.squareup.wire.ProtoWriter.varint32Size;
import static com.squareup.wire.ProtoWriter.varint64Size;
import static java.lang.Double.doubleToLongBits;
import static java.lang.Float.floatToIntBits;

public abstract class ProtoAdapter<E> {
  private static final int FIXED_BOOL_SIZE = 1;
  private static final int FIXED_32_SIZE = 4;
  private static final int FIXED_64_SIZE = 8;

  private final FieldEncoding fieldEncoding;
  final Class<?> javaType;

  ProtoAdapter<List<E>> packedAdapter;
  ProtoAdapter<List<E>> repeatedAdapter;

  public ProtoAdapter(FieldEncoding fieldEncoding, Class<?> javaType) {
    this.fieldEncoding = fieldEncoding;
    this.javaType = javaType;
  }

  /** Creates a new proto adapter for {@code type}. */
  public static <M extends Message<M, B>, B extends Builder<M, B>> ProtoAdapter<M>
      newMessageAdapter(Class<M> type) {
    return RuntimeMessageAdapter.create(type);
  }

  /** Creates a new proto adapter for {@code type}. */
  public static <E extends WireEnum> RuntimeEnumAdapter<E> newEnumAdapter(Class<E> type) {
    return new RuntimeEnumAdapter<>(type);
  }

  /** Returns the adapter for the type of {@code Message}. */
  public static <M extends Message> ProtoAdapter<M> get(M message) {
    return (ProtoAdapter<M>) get(message.getClass());
  }

  /** Returns the adapter for {@code type}. */
  public static <M> ProtoAdapter<M> get(Class<M> type) {
    try {
      return (ProtoAdapter<M>) type.getField("ADAPTER").get(null);
    } catch (IllegalAccessException | NoSuchFieldException e) {
      throw new IllegalArgumentException("failed to access " + type.getName() + "#ADAPTER", e);
    }
  }

  @SuppressWarnings("unchecked")
  static ProtoAdapter<?> get(String adapterString) {
    try {
      int hash = adapterString.indexOf('#');
      String className = adapterString.substring(0, hash);
      String fieldName = adapterString.substring(hash + 1);
      return (ProtoAdapter<Object>) Class.forName(className).getField(fieldName).get(null);
    } catch (IllegalAccessException | NoSuchFieldException | ClassNotFoundException e) {
      throw new IllegalArgumentException("failed to access " + adapterString, e);
    }
  }

  /** Returns the redacted form of {@code value}. */
  public E redact(E value) {
    return null;
  }

  /**
   * The size of the non-null data {@code value}. This does not include the size required for
   * a length-delimited prefix (should the type require one).
   */
  public abstract int encodedSize(E value);

  /**
   * The size of {@code tag} and non-null {@code value} in the wire format. This size includes the
   * tag, type, length-delimited prefix (should the type require one), and value.
   */
  public int encodedSizeWithTag(int tag, E value) {
    int size = encodedSize(value);
    if (fieldEncoding == FieldEncoding.LENGTH_DELIMITED) {
      size += varint32Size(size);
    }
    return size + ProtoWriter.tagSize(tag);
  }

  /** Write non-null {@code value} to {@code writer}. */
  public abstract void encode(ProtoWriter writer, E value) throws IOException;

  /** Write {@code tag} and non-null {@code value} to {@code writer}. */
  public void encodeWithTag(ProtoWriter writer, int tag, E value) throws IOException {
    writer.writeTag(tag, fieldEncoding);
    if (fieldEncoding == FieldEncoding.LENGTH_DELIMITED) {
      writer.writeVarint32(encodedSize(value));
    }
    encode(writer, value);
  }

  /** Encode {@code value} and write it to {@code stream}. */
  public final void encode(BufferedSink sink, E value) throws IOException {
    checkNotNull(value, "value == null");
    checkNotNull(sink, "sink == null");
    encode(new ProtoWriter(sink), value);
  }

  /** Encode {@code value} as a {@code byte[]}. */
  public final byte[] encode(E value) {
    checkNotNull(value, "value == null");
    Buffer buffer = new Buffer();
    try {
      encode(buffer, value);
    } catch (IOException e) {
      throw new AssertionError(e); // No I/O writing to Buffer.
    }
    return buffer.readByteArray();
  }

  /** Encode {@code value} and write it to {@code stream}. */
  public final void encode(OutputStream stream, E value) throws IOException {
    checkNotNull(value, "value == null");
    checkNotNull(stream, "stream == null");
    BufferedSink buffer = Okio.buffer(Okio.sink(stream));
    encode(buffer, value);
    buffer.emit();
  }

  /** Read a non-null value from {@code reader}. */
  public abstract E decode(ProtoReader reader) throws IOException;

  /** Read an encoded message from {@code bytes}. */
  public final E decode(byte[] bytes) throws IOException {
    checkNotNull(bytes, "bytes == null");
    return decode(new Buffer().write(bytes));
  }

  /** Read an encoded message from {@code bytes}. */
  public final E decode(ByteString bytes) throws IOException {
    checkNotNull(bytes, "bytes == null");
    return decode(new Buffer().write(bytes));
  }

  /** Read an encoded message from {@code stream}. */
  public final E decode(InputStream stream) throws IOException {
    checkNotNull(stream, "stream == null");
    return decode(Okio.buffer(Okio.source(stream)));
  }

  /** Read an encoded message from {@code source}. */
  public final E decode(BufferedSource source) throws IOException {
    checkNotNull(source, "source == null");
    return decode(new ProtoReader(source));
  }

  /** Returns a human-readable version of the given {@code value}. */
  public String toString(E value) {
    return value.toString();
  }

  public static final ProtoAdapter<Boolean> BOOL = new ProtoAdapter<Boolean>(
      FieldEncoding.VARINT, Boolean.class) {
    @Override public int encodedSize(Boolean value) {
      return FIXED_BOOL_SIZE;
    }

    @Override public void encode(ProtoWriter writer, Boolean value) throws IOException {
      writer.writeVarint32(value ? 1 : 0);
    }

    @Override public Boolean decode(ProtoReader reader) throws IOException {
      int value = reader.readVarint32();
      if (value == 0) return Boolean.FALSE;
      if (value == 1) return Boolean.TRUE;
      throw new IOException(String.format("Invalid boolean value 0x%02x", value));
    }
  };
  public static final ProtoAdapter<Integer> INT32 = new ProtoAdapter<Integer>(
      FieldEncoding.VARINT, Integer.class) {
    @Override public int encodedSize(Integer value) {
      return int32Size(value);
    }

    @Override public void encode(ProtoWriter writer, Integer value) throws IOException {
      writer.writeSignedVarint32(value);
    }

    @Override public Integer decode(ProtoReader reader) throws IOException {
      return reader.readVarint32();
    }
  };
  public static final ProtoAdapter<Integer> UINT32 = new ProtoAdapter<Integer>(
      FieldEncoding.VARINT, Integer.class) {
    @Override public int encodedSize(Integer value) {
      return varint32Size(value);
    }

    @Override public void encode(ProtoWriter writer, Integer value) throws IOException {
      writer.writeVarint32(value);
    }

    @Override public Integer decode(ProtoReader reader) throws IOException {
      return reader.readVarint32();
    }
  };
  public static final ProtoAdapter<Integer> SINT32 = new ProtoAdapter<Integer>(
      FieldEncoding.VARINT, Integer.class) {
    @Override public int encodedSize(Integer value) {
      return varint32Size(encodeZigZag32(value));
    }

    @Override public void encode(ProtoWriter writer, Integer value) throws IOException {
      writer.writeVarint32(encodeZigZag32(value));
    }

    @Override public Integer decode(ProtoReader reader) throws IOException {
      return decodeZigZag32(reader.readVarint32());
    }
  };
  public static final ProtoAdapter<Integer> FIXED32 = new ProtoAdapter<Integer>(
      FieldEncoding.FIXED32, Integer.class) {
    @Override public int encodedSize(Integer value) {
      return FIXED_32_SIZE;
    }

    @Override public void encode(ProtoWriter writer, Integer value) throws IOException {
      writer.writeFixed32(value);
    }

    @Override public Integer decode(ProtoReader reader) throws IOException {
      return reader.readFixed32();
    }
  };
  public static final ProtoAdapter<Integer> SFIXED32 = FIXED32;
  public static final ProtoAdapter<Long> INT64 = new ProtoAdapter<Long>(
      FieldEncoding.VARINT, Long.class) {
    @Override public int encodedSize(Long value) {
      return varint64Size(value);
    }

    @Override public void encode(ProtoWriter writer, Long value) throws IOException {
      writer.writeVarint64(value);
    }

    @Override public Long decode(ProtoReader reader) throws IOException {
      return reader.readVarint64();
    }
  };
  /**
   * Like INT64, but negative longs are interpreted as large positive values, and encoded that way
   * in JSON.
   */
  public static final ProtoAdapter<Long> UINT64 = new ProtoAdapter<Long>(
      FieldEncoding.VARINT, Long.class) {
    @Override public int encodedSize(Long value) {
      return varint64Size(value);
    }

    @Override public void encode(ProtoWriter writer, Long value) throws IOException {
      writer.writeVarint64(value);
    }

    @Override public Long decode(ProtoReader reader) throws IOException {
      return reader.readVarint64();
    }
  };
  public static final ProtoAdapter<Long> SINT64 = new ProtoAdapter<Long>(
      FieldEncoding.VARINT, Long.class) {
    @Override public int encodedSize(Long value) {
      return varint64Size(encodeZigZag64(value));
    }

    @Override public void encode(ProtoWriter writer, Long value) throws IOException {
      writer.writeVarint64(encodeZigZag64(value));
    }

    @Override public Long decode(ProtoReader reader) throws IOException {
      return decodeZigZag64(reader.readVarint64());
    }
  };
  public static final ProtoAdapter<Long> FIXED64 = new ProtoAdapter<Long>(
      FieldEncoding.FIXED64, Long.class) {
    @Override public int encodedSize(Long value) {
      return FIXED_64_SIZE;
    }

    @Override public void encode(ProtoWriter writer, Long value) throws IOException {
      writer.writeFixed64(value);
    }

    @Override public Long decode(ProtoReader reader) throws IOException {
      return reader.readFixed64();
    }
  };
  public static final ProtoAdapter<Long> SFIXED64 = FIXED64;
  public static final ProtoAdapter<Float> FLOAT = new ProtoAdapter<Float>(
      FieldEncoding.FIXED32, Float.class) {
    @Override public int encodedSize(Float value) {
      return FIXED_32_SIZE;
    }

    @Override public void encode(ProtoWriter writer, Float value) throws IOException {
      writer.writeFixed32(floatToIntBits(value));
    }

    @Override public Float decode(ProtoReader reader) throws IOException {
      return Float.intBitsToFloat(reader.readFixed32());
    }
  };
  public static final ProtoAdapter<Double> DOUBLE = new ProtoAdapter<Double>(
      FieldEncoding.FIXED64, Double.class) {
    @Override public int encodedSize(Double value) {
      return FIXED_64_SIZE;
    }

    @Override public void encode(ProtoWriter writer, Double value) throws IOException {
      writer.writeFixed64(doubleToLongBits(value));
    }

    @Override public Double decode(ProtoReader reader) throws IOException {
      return Double.longBitsToDouble(reader.readFixed64());
    }
  };
  public static final ProtoAdapter<String> STRING = new ProtoAdapter<String>(
      FieldEncoding.LENGTH_DELIMITED, String.class) {
    @Override public int encodedSize(String value) {
      return utf8Length(value);
    }

    @Override public void encode(ProtoWriter writer, String value) throws IOException {
      writer.writeString(value);
    }

    @Override public String decode(ProtoReader reader) throws IOException {
      return reader.readString();
    }
  };
  public static final ProtoAdapter<ByteString> BYTES = new ProtoAdapter<ByteString>(
      FieldEncoding.LENGTH_DELIMITED, ByteString.class) {
    @Override public int encodedSize(ByteString value) {
      return value.size();
    }

    @Override public void encode(ProtoWriter writer, ByteString value) throws IOException {
      writer.writeBytes(value);
    }

    @Override public ByteString decode(ProtoReader reader) throws IOException {
      return reader.readBytes();
    }
  };

  ProtoAdapter<?> withLabel(WireField.Label label) {
    if (label.isRepeated()) {
      return label.isPacked()
          ? asPacked()
          : asRepeated();
    }
    return this;
  }

  /** Returns an adapter for {@code E} but as a packed, repeated value. */
  public final ProtoAdapter<List<E>> asPacked() {
    ProtoAdapter<List<E>> adapter = packedAdapter;
    return adapter != null ? adapter : (packedAdapter = createPacked());
  }

  /** Returns an adapter for {@code E} but as a repeated value. */
  public final ProtoAdapter<List<E>> asRepeated() {
    ProtoAdapter<List<E>> adapter = repeatedAdapter;
    return adapter != null ? adapter : (repeatedAdapter = createRepeated());
  }

  private ProtoAdapter<List<E>> createPacked() {
    if (fieldEncoding == FieldEncoding.LENGTH_DELIMITED) {
      throw new IllegalArgumentException("Unable to pack a length-delimited type.");
    }
    return new ProtoAdapter<List<E>>(FieldEncoding.LENGTH_DELIMITED, List.class) {
      @Override public void encodeWithTag(ProtoWriter writer, int tag, List<E> value)
          throws IOException {
        if (!value.isEmpty()) {
          super.encodeWithTag(writer, tag, value);
        }
      }

      @Override public int encodedSize(List<E> value) {
        int size = 0;
        for (int i = 0, count = value.size(); i < count; i++) {
          size += ProtoAdapter.this.encodedSize(value.get(i));
        }
        return size;
      }

      @Override public int encodedSizeWithTag(int tag, List<E> value) {
        return value.isEmpty() ? 0 : super.encodedSizeWithTag(tag, value);
      }

      @Override public void encode(ProtoWriter writer, List<E> value) throws IOException {
        for (int i = 0, count = value.size(); i < count; i++) {
          ProtoAdapter.this.encode(writer, value.get(i));
        }
      }

      @Override public List<E> decode(ProtoReader reader) throws IOException {
        E value = ProtoAdapter.this.decode(reader);
        return Collections.singletonList(value);
      }

      @Override public List<E> redact(List<E> value) {
        return Collections.emptyList();
      }
    };
  }

  private ProtoAdapter<List<E>> createRepeated() {
    return new ProtoAdapter<List<E>>(fieldEncoding, List.class) {
      @Override public int encodedSize(List<E> value) {
        throw new UnsupportedOperationException("Repeated values can only be sized with a tag.");
      }

      @Override public int encodedSizeWithTag(int tag, List<E> value) {
        int size = 0;
        for (int i = 0, count = value.size(); i < count; i++) {
          size += ProtoAdapter.this.encodedSizeWithTag(tag, value.get(i));
        }
        return size;
      }

      @Override public void encode(ProtoWriter writer, List<E> value) {
        throw new UnsupportedOperationException("Repeated values can only be encoded with a tag.");
      }

      @Override public void encodeWithTag(ProtoWriter writer, int tag, List<E> value)
          throws IOException {
        for (int i = 0, count = value.size(); i < count; i++) {
          ProtoAdapter.this.encodeWithTag(writer, tag, value.get(i));
        }
      }

      @Override public List<E> decode(ProtoReader reader) throws IOException {
        E value = ProtoAdapter.this.decode(reader);
        return Collections.singletonList(value);
      }

      @Override public List<E> redact(List<E> value) {
        return Collections.emptyList();
      }
    };
  }

  public static final class EnumConstantNotFoundException extends IllegalArgumentException {
    public final int value;

    EnumConstantNotFoundException(int value, Class<?> type) {
      super("Unknown enum tag " + value + " for " + type.getCanonicalName());
      this.value = value;
    }
  }
}

File: wire-runtime/src/nativeMain/kotlin/com/squareup/wire/ProtoAdapter.kt
/*
 * Copyright 2015 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire

import okio.BufferedSink
import okio.BufferedSource
import okio.ByteString
import kotlin.reflect.KClass

actual abstract class ProtoAdapter<E> actual constructor(
  internal actual val fieldEncoding: FieldEncoding,
  actual val type: KClass<*>?,
  actual val typeUrl: String?
) {
  internal actual val packedAdapter: ProtoAdapter<List<E>>? = when {
    this is PackedProtoAdapter<*> || this is RepeatedProtoAdapter<*> -> null
    fieldEncoding == FieldEncoding.LENGTH_DELIMITED -> null
    else -> commonCreatePacked()
  }
  internal actual val repeatedAdapter: ProtoAdapter<List<E>>? = when {
    this is RepeatedProtoAdapter<*> || this is PackedProtoAdapter<*> -> null
    else -> commonCreateRepeated()
  }

  /** Returns the redacted form of `value`. */
  actual abstract fun redact(value: E): E

  /**
   * The size of the non-null data `value`. This does not include the size required for a
   * length-delimited prefix (should the type require one).
   */
  actual abstract fun encodedSize(value: E): Int

  /**
   * The size of `tag` and `value` in the wire format. This size includes the tag, type,
   * length-delimited prefix (should the type require one), and value. Returns 0 if `value` is
   * null.
   */
  actual open fun encodedSizeWithTag(tag: Int, value: E?): Int {
    return commonEncodedSizeWithTag(tag, value)
  }

  /** Write non-null `value` to `writer`. */
  actual abstract fun encode(writer: ProtoWriter, value: E)

  /** Write `tag` and `value` to `writer`. If value is null this does nothing. */
  actual open fun encodeWithTag(writer: ProtoWriter, tag: Int, value: E?) {
    commonEncodeWithTag(writer, tag, value)
  }

  /** Encode `value` and write it to `stream`. */
  actual fun encode(sink: BufferedSink, value: E) {
    commonEncode(sink, value)
  }

  /** Encode `value` as a `byte[]`. */
  actual fun encode(value: E): ByteArray {
    return commonEncode(value)
  }

  /** Encode `value` as a [ByteString]. */
  actual fun encodeByteString(value: E): ByteString {
    return commonEncodeByteString(value)
  }

  /** Read a non-null value from `reader`. */
  actual abstract fun decode(reader: ProtoReader): E

  /** Read an encoded message from `bytes`. */
  actual fun decode(bytes: ByteArray): E {
    return commonDecode(bytes)
  }

  /** Read an encoded message from `bytes`. */
  actual fun decode(bytes: ByteString): E {
    return commonDecode(bytes)
  }

  /** Read an encoded message from `source`. */
  actual fun decode(source: BufferedSource): E {
    return commonDecode(source)
  }

  /** Returns a human-readable version of the given `value`. */
  actual open fun toString(value: E): String {
    return commonToString(value)
  }

  internal actual fun withLabel(label: WireField.Label): ProtoAdapter<*> {
    return commonWithLabel(label)
  }

  /** Returns an adapter for `E` but as a packed, repeated value. */
  actual fun asPacked(): ProtoAdapter<List<E>> {
    require(fieldEncoding != FieldEncoding.LENGTH_DELIMITED) {
      "Unable to pack a length-delimited type."
    }
    return packedAdapter ?: throw UnsupportedOperationException(
        "Can't create a packed adapter from a packed or repeated adapter.")
  }

  /**
   * Returns an adapter for `E` but as a repeated value.
   *
   * Note: Repeated items are not required to be encoded sequentially. Thus, when decoding using
   * the returned adapter, only single-element lists will be returned and it is the caller's
   * responsibility to merge them into the final list.
   */
  actual fun asRepeated(): ProtoAdapter<List<E>> {
    return repeatedAdapter ?: throw UnsupportedOperationException(
        "Can't create a repeated adapter from a repeated or packed adapter.")
  }

  actual class EnumConstantNotFoundException actual constructor(
    actual val value: Int,
    type: KClass<*>?
  ) : IllegalArgumentException("Unknown enum tag $value for ${type?.simpleName}")

  actual companion object {
    /**
     * Creates a new proto adapter for a map using `keyAdapter` and `valueAdapter`.
     *
     * Note: Map entries are not required to be encoded sequentially. Thus, when decoding using
     * the returned adapter, only single-element maps will be returned and it is the caller's
     * responsibility to merge them into the final map.
     */
    actual fun <K, V> newMapAdapter(
      keyAdapter: ProtoAdapter<K>,
      valueAdapter: ProtoAdapter<V>
    ): ProtoAdapter<Map<K, V>> {
      return commonNewMapAdapter(keyAdapter, valueAdapter)
    }

    actual val BOOL: ProtoAdapter<Boolean> = commonBool()
    actual val INT32: ProtoAdapter<Int> = commonInt32()
    actual val UINT32: ProtoAdapter<Int> = commonUint32()
    actual val SINT32: ProtoAdapter<Int> = commonSint32()
    actual val FIXED32: ProtoAdapter<Int> = commonFixed32()
    actual val SFIXED32: ProtoAdapter<Int> = commonSfixed32()
    actual val INT64: ProtoAdapter<Long> = commonInt64()
    /**
     * Like INT64, but negative longs are interpreted as large positive values, and encoded that way
     * in JSON.
     */
    actual val UINT64: ProtoAdapter<Long> = commonUint64()
    actual val SINT64: ProtoAdapter<Long> = commonSint64()
    actual val FIXED64: ProtoAdapter<Long> = commonFixed64()
    actual val SFIXED64: ProtoAdapter<Long> = commonSfixed64()
    actual val FLOAT: ProtoAdapter<Float> = commonFloat()
    actual val DOUBLE: ProtoAdapter<Double> = commonDouble()
    actual val BYTES: ProtoAdapter<ByteString> = commonBytes()
    actual val STRING: ProtoAdapter<String> = commonString()
  }
}
